### PR TITLE
Kaa-1143. CTL in Configuration management extension.

### PR DESCRIPTION
**Pull request by Kirill380**

### CODE REVIEW

File: common/core/src/main/java/org/kaaproject/kaa/server/common/core/algorithms/AvroUtils.java [hunk 1]
```diff
@@ -92,4 +92,38 @@ public static void copyJsonProperties(JsonProperties src, JsonProperties dst) {
...
+        boolean notArray = !json.isArray();
+        boolean childIsNotArray = !(json.size() == 1 && json.getElements().next() instanceof ArrayNode);
+
+        if (containerWithoutId && notArray && childIsNotArray) {
+            ((ObjectNode)json).put(UUID_FIELD, (Integer)null);
+        }
+
+        for (JsonNode node : json) {
+            if (node.isContainerNode())
+                injectUuids(node);
```
Comment (ashvayka) [inline common/core/src/main/java/org/kaaproject/kaa/server/common/core/algorithms/AvroUtils.java:107]: I am not sure that this is method has correct API. Since you ignore the result of the method here, I believe it is better to return void. This will clarify behaviour of the method to the users.

File: common/core/src/main/java/org/kaaproject/kaa/server/common/core/algorithms/AvroUtils.java [hunk 2]
```diff
@@ -92,4 +92,38 @@ public static void copyJsonProperties(JsonProperties src, JsonProperties dst) {
...
+        }
+
+        return json;
+    }
+
+    public static JsonNode removeUuids(JsonNode json) {
```
Comment (ashvayka) [inline common/core/src/main/java/org/kaaproject/kaa/server/common/core/algorithms/AvroUtils.java]: same comment here.

File: common/endpoint-shared/src/main/java/org/kaaproject/kaa/common/Constants.java [hunk 3]
```diff
@@ -20,44 +20,38 @@
  * Common Kaa project Constants.
  */
 public interface Constants { //NOSONAR
-    /**
-     * Used URI delimiter.
-     */
-    public static final String URI_DELIM = "/"; //NOSONAR
+
+    String URI_DELIM = "/"; //NOSONAR
```
Comment (ashvayka) [inline common/endpoint-shared/src/main/java/org/kaaproject/kaa/common/Constants.java:24]: Why you have applied this change?

Comment (ashvayka) [inline common/endpoint-shared/src/main/java/org/kaaproject/kaa/common/Constants.java:24]: Maybe convert Constants to the final class?
See http://stackoverflow.com/questions/2659593/what-is-the-use-of-interface-constants

File: common/endpoint-shared/src/main/java/org/kaaproject/kaa/common/channels/protocols/kaatcp/messages/MessageFactory.java [hunk 4]
```diff
@@ -81,154 +72,109 @@ public MessageFactory() {
...
-            }
-            break;
-        case PINGRESP:
-            if (pingResponseListener != null) {
-                pingResponseListener.onMessage((PingResponse) frame);
-            }
-            break;
-        default:
-            break;
+            case CONNACK:
```
Comment (ashvayka) [inline common/endpoint-shared/src/main/java/org/kaaproject/kaa/common/channels/protocols/kaatcp/messages/MessageFactory.java]: Please try to avoid such "formatting" changes. This increase complexity of your PR and makes us spend more time on review.

File: server/common/dao/src/main/java/org/kaaproject/kaa/server/common/dao/EndpointRegistrationService.java [hunk 5]
```diff
@@ -1,17 +1,17 @@
-/**
- *  Copyright 2014-2016 CyberVision, Inc.
+/*
```
Comment (ashvayka) [inline server/common/dao/src/main/java/org/kaaproject/kaa/server/common/dao/EndpointRegistrationService.java]: Please remove all unrelated changes from this PR.

File: server/common/nosql/cassandra-dao/src/main/resources/cassandra.cql [hunk 6]
```diff
@@ -88,6 +88,7 @@ CREATE TABLE IF NOT EXISTS kaa.ep_profile (
     server_hash text,
     sdk_token text,
     srv_pf text,
+    use_raw_schema boolean,
```
Comment (ashvayka) [inline server/common/nosql/cassandra-dao/src/main/resources/cassandra.cql:91]: Need to provide upgrade script for this change

File: server/common/server-shared/src/main/java/org/kaaproject/kaa/server/sync/platform/AvroEncDecUseRawChema.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.kaaproject.kaa.common.Constants;
+import org.kaaproject.kaa.common.endpoint.gen.ProfileSyncRequest;
+import org.kaaproject.kaa.common.endpoint.gen.SyncRequest;
+import org.kaaproject.kaa.server.sync.ClientSync;
+import org.kaaproject.kaa.server.sync.ProfileClientSync;
+import org.kaaproject.kaa.server.sync.platform.AvroEncDec;
+
+@KaaPlatformProtocol
+public class AvroEncDecUseRawChema extends AvroEncDec {
```
Comment (ashvayka) [inline server/common/server-shared/src/main/java/org/kaaproject/kaa/server/sync/platform/AvroEncDecUseRawChema.java]: RawConfigurationSchemaAvroEncDec

File: server/common/server-shared/src/main/java/org/kaaproject/kaa/server/sync/platform/BinaryEncDecUseRawSchema.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+package org.kaaproject.kaa.server.sync.platform;
+
+
+import org.kaaproject.kaa.common.Constants;
+
+import org.kaaproject.kaa.server.sync.ClientSync;
+
+@KaaPlatformProtocol
+public class BinaryEncDecUseRawSchema extends BinaryEncDec{
```
Comment (ashvayka) [inline server/common/server-shared/src/main/java/org/kaaproject/kaa/server/sync/platform/BinaryEncDecUseRawSchema.java]: RawConfigurationSchemaBinaryEncDec

Comment (ashvayka) [inline server/common/server-shared/src/main/java/org/kaaproject/kaa/server/sync/platform/BinaryEncDecUseRawSchema.java]: whitespace "extends BinaryEncDec {"

File: server/node/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/activity/CtlSchemaActivity.java [hunk 9]
```diff
@@ -160,8 +154,16 @@ public void onFailureImpl(Throwable caught) {
     @Override
     protected void onEntityRetrieved() {
         if (create) {
-            KaaAdmin.getDataSource().createNewCTLSchemaFormInstance(place.getMetaInfoId(), 
-                    place.getVersion(), place.getApplicationId(), 
+            ConverterType converterType;
+
+            if(place.getSchemaType() == SchemaType.CONFIGURATION) {
```
Comment (ashvayka) [inline server/node/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/activity/CtlSchemaActivity.java]: spacing

File: server/node/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/place/CtlSchemaPlace.java [hunk 10]
```diff
@@ -182,6 +182,8 @@ public TreePlace createDefaultPreviousPlace() {
             if (schemaType != null) { 
                 if (schemaType == SchemaType.ENDPOINT_PROFILE) {
                     return new ProfileSchemasPlace(applicationId);
+                } else if(schemaType == SchemaType.CONFIGURATION) {
```
Comment (ashvayka) [inline server/node/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/place/CtlSchemaPlace.java]: spacing

File: server/node/src/main/java/org/kaaproject/kaa/server/admin/services/KaaAdminServiceImpl.java [hunk 11]
```diff
@@ -961,11 +964,20 @@ public RecordField createCommonEmptySchemaForm() throws KaaAdminServiceException
...
+
     @Override
-    public RecordField createConfigurationEmptySchemaForm() throws KaaAdminServiceException {
+    public ConfigurationSchemaViewDto createConfigurationSchemaFormCtlSchema(CtlSchemaFormDto ctlSchemaForm) throws KaaAdminServiceException {
         checkAuthority(KaaAuthorityDto.TENANT_ADMIN, KaaAuthorityDto.TENANT_DEVELOPER, KaaAuthorityDto.TENANT_USER);
         try {
-            return configurationSchemaFormAvroConverter.getEmptySchemaFormInstance();
+            checkApplicationId(ctlSchemaForm.getMetaInfo().getApplicationId());
+            ConfigurationSchemaDto confSchema = new ConfigurationSchemaDto();
+            confSchema.setApplicationId(ctlSchemaForm.getMetaInfo().getApplicationId());
```
Comment (ashvayka) [inline server/node/src/main/java/org/kaaproject/kaa/server/admin/services/KaaAdminServiceImpl.java]: Sometimes application id may be null, right?

**Comment by ashvayka** [comment]
I am done with initial review. Please fix your PR according to my comments. I will review it once again after your fixes.

File: server/node/src/main/java/org/kaaproject/kaa/server/admin/controller/KaaAdminController.java [hunk 12]
```diff
@@ -1070,37 +1070,16 @@ public ConfigurationSchemaDto getConfigurationSchema(@PathVariable String config
...
+     * @throws KaaAdminServiceException the kaa admin service exception
      */
-    @RequestMapping(value = "createConfigurationSchema", method = RequestMethod.POST, consumes = { "multipart/mixed", "multipart/form-data" })
+    @RequestMapping(value = "saveConfigurationSchema", method = RequestMethod.POST)
     @ResponseBody
-    public ConfigurationSchemaDto createConfigurationSchema(@RequestPart("configurationSchema") ConfigurationSchemaDto configurationSchema,
-            @RequestPart("file") MultipartFile file) throws KaaAdminServiceException {
-        byte[] data = getFileContent(file);
-        return kaaAdminService.editConfigurationSchema(configurationSchema, data);
+    public ConfigurationSchemaDto createConfigurationSchema(@RequestBody  ConfigurationSchemaDto configurationSchema) throws KaaAdminServiceException {
```
Comment (ikulikov) [inline server/node/src/main/java/org/kaaproject/kaa/server/admin/controller/KaaAdminController.java]: Please rename to saveConfigurationSchema

File: server/node/src/main/java/org/kaaproject/kaa/server/operations/service/delta/DefaultDeltaService.java [hunk 13]
```diff
@@ -208,12 +213,19 @@ public ConfigurationCacheEntry compute(DeltaCacheKey deltaKey) {
                                 userConfigurationHash = EndpointObjectHash.fromString(userConfiguration.getBody());
                             }
 
-                            BaseData mergedConfiguration = getMergedConfiguration(endpointId, userConfiguration, deltaKey,
-                                    latestConfigurationSchema);
+                            BaseData mergedConfiguration = getMergedConfiguration(endpointId, userConfiguration, deltaKey, latestConfigurationSchema);
 
-                            LOG.trace("[{}] Merged configuration {}", endpointId, mergedConfiguration.getRawData());
+                            if(useConfigurationRawSchema) {
+                                kaaData = mergedConfiguration;
```
Comment (ikulikov) [inline server/node/src/main/java/org/kaaproject/kaa/server/operations/service/delta/DefaultDeltaService.java]: Raw configuration schema must be used here.

File: server/node/src/main/java/org/kaaproject/kaa/server/operations/service/delta/DefaultDeltaService.java [hunk 14]
```diff
@@ -351,10 +363,12 @@ public BaseData compute(List<EndpointGroupStateDto> key) {
         return mergedConfiguration;
     }
 
-    private ConfigurationCacheEntry buildBaseResyncDelta(String endpointId, BaseData mergedConfiguration, EndpointObjectHash userConfigurationHash) {
-        byte[] configuration = GenericAvroConverter.toRawData(mergedConfiguration.getRawData(), mergedConfiguration.getSchema()
-                .getRawSchema());
+    private ConfigurationCacheEntry buildBaseResyncDelta(String endpointId, AbstractKaaData<?> mergedConf, EndpointObjectHash userConfigurationHash) throws IOException {
+        JsonNode json = new ObjectMapper().readTree(mergedConf.getRawData());
+        AvroUtils.removeUuids(json);
```
Comment (ikulikov) [inline server/node/src/main/java/org/kaaproject/kaa/server/operations/service/delta/DefaultDeltaService.java]: Separate processing of raw and base configuration records according to the 'useConfigurationRawSchema' parameter.

File: server/node/src/main/java/org/kaaproject/kaa/server/operations/service/delta/DefaultDeltaService.java [hunk 15]
```diff
@@ -208,12 +213,19 @@ public ConfigurationCacheEntry compute(DeltaCacheKey deltaKey) {
...
-                            BaseData mergedConfiguration = getMergedConfiguration(endpointId, userConfiguration, deltaKey,
-                                    latestConfigurationSchema);
+                            BaseData mergedConfiguration = getMergedConfiguration(endpointId, userConfiguration, deltaKey, latestConfigurationSchema);
 
-                            LOG.trace("[{}] Merged configuration {}", endpointId, mergedConfiguration.getRawData());
+                            if(useConfigurationRawSchema) {
+                                kaaData = mergedConfiguration;
+                            } else {
+                                // converting merged base schema to raw schema
+                                String ctlSchema = cacheService.getFlatCtlSchemaById(latestConfigurationSchema.getCtlSchemaId());
```
Comment (ashvayka) [inline server/node/src/main/java/org/kaaproject/kaa/server/operations/service/delta/DefaultDeltaService.java]: This operation need to be cached. I suggest to modify "getMergedConfiguration" method to return simple structure instead of BaseData. This structure will contain both BaseData and RawData

**Comment by ashvayka** [comment]
LGTM

**Comment by ikulikov** [comment]
LGTM
